src/coprocessor/trie/mod.rs

```diff
@@ -0,0 +1,734 @@
+            6 => hash!(hash6, 6),
+            8 => hash!(hash8, 8),
+            _ => unimplemented!(),
+        }
+    }
+
+    pub fn root(&self) -> F {
+        self.root
+    }
+
+    /// How many leaves does this `Trie` have?
+    pub const fn leaves(&self) -> usize {
+        self.row_size(HEIGHT)
+    }
+
+    /// How many nodes does the `row`th row have?
+    pub const fn row_size(&self, row: usize) -> usize {
+        debug_assert!(row <= HEIGHT);
+        (ARITY as u32).pow(row as u32) as usize
+    }
+
+    fn register_hash(&mut self, preimage: [F; ARITY]) -> F {
+        let hash = Self::compute_hash(self.hash_cache, preimage);
+
+        self.children.insert(FWrap(hash), preimage);
+        hash
+    }
+
+    fn get_hash_preimage(children: &ChildMap<F, ARITY>, hash: F) -> Option<&[F; ARITY]> {
+        children.get(&FWrap(hash))
+    }
+
+    // TODO: This is expensive and can be cached for the type.
+    fn init_empty(&mut self) {
+        self.empty_roots = [F::zero(); HEIGHT];
+
+        if HEIGHT == 0 {
+            return;
+        };
+
+        let mut preimage = [self.empty_roots[0]; ARITY];
+
+        for i in 0..HEIGHT {
+            let hash = self.register_hash(preimage);
+
+            self.empty_roots[i] = hash;
+
+            preimage = [hash; ARITY];
+            for elt in preimage.iter_mut().take(ARITY) {
+                *elt = hash;
+            }
+        }
+        self.count = 0;
+        self.root = self.empty_roots[HEIGHT - 1]
+    }
+
+    fn empty_root_for_height(&self, height: usize) -> F {
+        self.empty_roots[height - 1]
+    }
+
+    pub fn empty_root(&mut self) -> F {
+        self.empty_root_for_height(HEIGHT)
+    }
+    fn new(
+        poseidon_cache: &'a PoseidonCache<F>,
+        inverse_poseidon_cache: &'a mut InversePoseidonCache<F>,
+    ) -> Self {
+        // ARITY must be a power of two.
+        assert_eq!(1, ARITY.count_ones());
+
+        let mut new = Self {
+            root: Default::default(),
+            empty_roots: [F::zero(); HEIGHT],
+            hash_cache: poseidon_cache,
+            children: inverse_poseidon_cache,
+            count: 0,
+        };
+        new.init_empty();
+
+        new
+    }
+
+    /// Create a new `Trie` with specified root. The provided `PoseidonCache` must
+    fn new_with_root(store: &'a mut Store<F>, root: F) -> Self {
+        let poseidon_cache = &store.poseidon_cache;
+        let inverse_poseidon_cache = &mut store.inverse_poseidon_cache;
+        let mut new = Self::new(poseidon_cache, inverse_poseidon_cache);
+        // FIXME: count is not initialized.
+        new.root = root;
+
+        new
+    }
+
+    const fn arity_bits() -> usize {
+        ARITY.trailing_zeros() as usize
+    }
+
+    /// (arity_bits, path_bits)
+    const fn path_bit_dimensions() -> (usize, usize) {
+        let arity_bits = Self::arity_bits();
+        (arity_bits, arity_bits * HEIGHT)
+    }
+
+    fn path(key: F) -> Vec<usize> {
```

Comment:
Wouldn't working with a u64 here give us a more platform-independent logic?

Reply:
I don't think it matters. `usize` is just the natural type to use, but since the path elements are bounded by the branching factor of the tree, in practice they will be 'small' — so I don't think we would ever conceivably require values that are larger than the `usize` of any architecture.

Oct trees are optimal for neptune/poseidon inclusion proofs. Where tree structure is important to the application, we may want multidimensional tries requiring arity 2^n.